src/Compilers/Core/Portable/Compilation.EmitStream.cs

```diff
@@ -0,0 +1,176 @@
+        }
+
+        /// <summary>
+        /// This type abstracts away the legacy COM based signing implementation for PE streams. Under the hood
+        /// a temporary file must be created on disk (at the last possible moment), emitted to, signed on disk
+        /// and then copied back to the original <see cref="Stream"/>. Only when legacy signing is enabled though.
+        /// </summary>
+        internal sealed class EmitStream
+        {
+            private readonly EmitStreamProvider _emitStreamProvider;
+            private readonly EmitStreamSignKind _emitStreamSignKind;
+            private readonly StrongNameProvider _strongNameProvider;
+            private (Stream tempStream, string tempFilePath)? _tempInfo;
+
+            /// <summary>
+            /// The <see cref="Stream"/> that is being emitted into. This value should _never_ be 
+            /// Dispose. It is either returned from the <see cref="EmitStreamProvider"/> instance in
+            /// which case it is owned by that. Or it is just an alias for the value that is stored 
+            /// in <see cref="_tempInfo"/> in which case it will be disposed from there.
+            /// </summary>
+            private Stream _stream;
+
+            internal EmitStream(
+                EmitStreamProvider emitStreamProvider,
+                EmitStreamSignKind emitStreamSignKind,
+                StrongNameProvider strongNameProvider)
+            {
+                Debug.Assert(emitStreamProvider != null);
+                Debug.Assert(strongNameProvider != null || emitStreamSignKind == EmitStreamSignKind.None);
+                _emitStreamProvider = emitStreamProvider;
+                _emitStreamSignKind = emitStreamSignKind;
+                _strongNameProvider = strongNameProvider;
+            }
+
+            internal Func<Stream> GetCreateStreamFunc(DiagnosticBag diagnostics)
+            {
+                return () => CreateStream(diagnostics);
+            }
+
+            internal void Dispose()
+            {
+                _tempInfo?.tempStream.Dispose();
+                _tempInfo = null;
+
+                // The _stream value is deliberately excluded from being disposed here. That value is not 
+                // owned by this type.
+                _stream = null;
+            }
+
+            /// <summary>
+            /// Create the stream which should be used for Emit. This should only be called one time.
+            /// </summary>
+            private Stream CreateStream(DiagnosticBag diagnostics)
+            {
+                Debug.Assert(_stream == null);
+                Debug.Assert(diagnostics != null);
+
+                if (diagnostics.HasAnyErrors())
+                {
+                    return null;
+                }
+
+                _stream = _emitStreamProvider.GetOrCreateStream(diagnostics);
+                if (_stream == null)
+                {
+                    Debug.Assert(diagnostics.HasAnyErrors());
```

Comment:
This cannot possibly ever be true.

Reply:
Technically it can if `_emitStreamProvider` captured `diagnostics`. That's pretty pathological though. I didn't add this assert, just copied it from the original location. Agree it should be removed though.

---
In reply to: [245374544](https://github.com/dotnet/roslyn/pull/31986#discussion_r245374544) [](ancestors = 245374544)